test_engine/build_test_engine.py

```diff
@@ -0,0 +1,243 @@
+
+    version_tag1 = "{}-{}.{}-{}.{}-{}".format(
+        int(CLIAPIMinVersion) - 1, int(CLIAPIMinVersion) - 1,
+        ControllerAPIVersion, ControllerAPIMinVersion,
+        DataFormatVersion, DataFormatMinVersion)
+
+    version_tag2 = "{}-{}.{}-{}.{}-{}".format(
+        int(CLIAPIVersion) + 1, int(CLIAPIVersion) + 1,
+        ControllerAPIVersion, ControllerAPIMinVersion,
+        DataFormatVersion, DataFormatMinVersion)
+
+    upgrade_image = "{0}:upgrade-test.{1}".format(docker_repo, docker_tag)
+    version_image1 = "{0}:version-test.{1}".format(docker_repo, version_tag1)
+    version_image2 = "{0}:version-test.{1}".format(docker_repo, version_tag2)
+
+    print(upgrade_image)
+    print(version_image1)
+    print(version_image2)
+
+    return upgrade_image, version_image1, version_image2
+
+
+def __exec_command(ip, username, commands):
+
+    client = paramiko.SSHClient()
+    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    client.connect(ip, username=username, key_filename=os.path.expanduser(
+                    os.path.join("~", ".ssh", "id_rsa")), timeout=1500)
+
+    stdin, stdout, stderr = client.exec_command(commands)
+
+    while int(stdout.channel.recv_exit_status()) != 0:
+        time.sleep(RETRY_INTERVAL)
+
+    output = stdout.readlines()
+
+    client.close()
+
+    print(output)
+
+    return output
+
+
+def add_manifest(upgrade_image, version_image1, version_image2):
+
+    # upgrade image
+    commands = 'docker login -u={0} -p={1};\
+        docker manifest create {2} \\{2}-amd64 \\{2}-arm64;\
+        docker manifest push {2}'\
+            .format(docker_id, docker_pwd, upgrade_image)
+    print(commands)
+    subprocess.run(commands, shell=True)
+
+    # version image
+    commands = '\
+        docker manifest create {0} \\{0}-amd64 \\{0}-arm64;\
+        docker manifest push {0}'\
+            .format(version_image1)
+    print(commands)
+    subprocess.run(commands, shell=True)
+
+    # version image
+    commands = '\
+        docker manifest create {0} \\{0}-amd64 \\{0}-arm64;\
+        docker manifest push {0}'\
+            .format(version_image2)
+    print(commands)
+    subprocess.run(commands, shell=True)
+
+
+def generate_test_images():
+
+    # arch : amd64 arm64
+    upgrade_image, version_image1, version_image2 = get_image_names()
+    build_images("amd64", upgrade_image, version_image1, version_image2)
+    build_images("arm64", upgrade_image, version_image1, version_image2)
+    add_manifest(upgrade_image, version_image1, version_image2)
+
+
+def build_images(arch, upgrade_image, version_image1, version_image2):
+
+    os.environ['TF_VAR_build_engine_arch'] = arch
+
+    if arch == "amd64":
+
+        os.environ['TF_VAR_build_engine_aws_instance_type'] = 't2.xlarge'
+
+    elif arch == "arm64":
+
+        os.environ['TF_VAR_build_engine_aws_instance_type'] = 'a1.xlarge'
+
+    subprocess.check_output("{}/scripts/terraform_instance.sh".format(
```

Comment:
Copied from `longhorn-tests/test_framework/scripts/build-engine.sh` and relocated/modified/renamed to `longhorn-tests/test_framework/scripts/terraform_instance.sh` in case we can move folder `test_engine` to other place/repo without dependency